buildpacks/go/faas/main.go

```diff
@@ -33,23 +43,92 @@ func main() {
 
 // run a cloudevents client in receive mode which invokes
 // the user-defined function.Handler on receipt of an event.
-func run() (err error) {
-	transport, err := cloudevents.NewHTTP(
-		cloudevents.WithPort(*port),
-		cloudevents.WithPath("/"))
+func run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		cancel()
+	}()
+
+	var handler interface{} = function.Handle
+
+	httpHandler, err := toHttpHandler(handler, ctx)
 	if err != nil {
-		return
+		return err
 	}
-	client, err := cloudevents.NewClient(transport)
-	if err != nil {
-		return
+	if httpHandler != nil {
+		httpServer := &http.Server{
+			Addr:           fmt.Sprintf(":%d", *port),
+			Handler:        httpHandler,
+			ReadTimeout:    10 * time.Second,
+			WriteTimeout:   10 * time.Second,
+			MaxHeaderBytes: 1 << 20,
+		}
+		go func() {
+			if *verbose {
+				fmt.Printf("listening on http port %v for HTTP requests\n", *port)
+			}
+			err = httpServer.ListenAndServe()
+			if err != nil {
+				fmt.Println(err)
+			}
+		}()
+		<- ctx.Done()
+		shutdownCtx, _ := context.WithTimeout(context.Background(), time.Second * 10)
+		httpServer.Shutdown(shutdownCtx)
+	} else {
+		transport, err := cloudevents.NewHTTP(
+			cloudevents.WithPort(*port),
+			cloudevents.WithPath("/"))
+		if err != nil {
+			return err
+		}
+		client, err := cloudevents.NewClient(transport)
+		if err != nil {
+			return err
+		}
+		if *verbose {
+			fmt.Printf("listening on http port %v for JSON-encoded CloudEvents\n", *port)
+		}
+		err = client.StartReceiver(ctx, handler)
 	}
-	if *verbose {
-		fmt.Printf("listening on http port %v for JSON-encoded CloudEvents\n", *port)
+
+	return err
+}
+
+// if the handler signature is http handler compatible the function returns http.Handler,
+// otherwise nil is returned
+func toHttpHandler(handler interface{}, ctx context.Context) (http.Handler, error) {
+
+	fnType := reflect.TypeOf(handler)
+	if fnType.Kind() != reflect.Func {
+		return nil, errors.New("must pass a function to handle events/http")
+	}
+
+	if fnType.NumOut() > 0 {
+		return nil, nil
+	}
+
+	if fnType.In(0) == httpResponseWriterType && fnType.In(1) == httpRequestPtrType {
+		if f, ok := handler.(func(rw http.ResponseWriter, req *http.Request)); ok {
+			return http.HandlerFunc(f), nil
+		}
+	}
+	if fnType.In(0) == contextType &&
+		fnType.In(1) == httpResponseWriterType &&
+		fnType.In(2) == httpRequestPtrType {
+		if f, ok := handler.(func(ctx context.Context, rw http.ResponseWriter, req *http.Request)); ok {
```

Comment:
This type assert is probably sufficient and the reflective check is not necessary.